### PR TITLE
Rule: identical conditions in if/match

### PR DESCRIPTION
**Pull request by pynicolas**

### CODE REVIEW

File: slang-checks/src/main/java/com/sonarsource/slang/checks/utils/SyntacticEquivalence.java [hunk 1]
```diff
@@ -61,4 +66,45 @@ public static boolean areEquivalent(Tree first, Tree second) {
...
+    ComparableTree(Tree tree) {
+      this.tree = tree;
+      hash = computeHash(tree);
+    }
+
+    private static int computeHash(@Nullable Tree tree) {
+      if (tree == null) {
+        return 0;
+      }
+      return TreePrinter.tree2string(tree).hashCode();
```
Comment (christophe-zurn-sonarsource) [inline slang-checks/src/main/java/com/sonarsource/slang/checks/utils/SyntacticEquivalence.java]: `tree2string` method does not implement printing for new trees (`BlockTree`, `FunctionDeclarationTree`, `IfTree`, `MatchTree` and `MatchCaseTree`), so we may have various surprises there...

File: slang-checks/src/main/java/com/sonarsource/slang/checks/utils/SyntacticEquivalence.java [hunk 2]
```diff
@@ -61,4 +66,45 @@ public static boolean areEquivalent(Tree first, Tree second) {
 
     return true;
   }
+
+  public static List<List<Tree>> findDuplicatedGroups(List<Tree> list) {
+    return list.stream()
```
Comment (christophe-zurn-sonarsource) [inline slang-checks/src/main/java/com/sonarsource/slang/checks/utils/SyntacticEquivalence.java]: Can you please add a unit test for this method?

File: slang-checks/src/main/java/com/sonarsource/slang/checks/utils/SyntacticEquivalence.java [hunk 3]
```diff
@@ -24,6 +24,11 @@
 import com.sonarsource.slang.api.LiteralTree;
 import com.sonarsource.slang.api.NativeTree;
 import com.sonarsource.slang.api.Tree;
+import com.sonarsource.slang.visitors.TreePrinter;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
 
 public class SyntacticEquivalence {
```
Comment (christophe-zurn-sonarsource) [inline slang-checks/src/main/java/com/sonarsource/slang/checks/utils/SyntacticEquivalence.java]: The `areEquivalent` method should also be updated accordingly with the new types of Tree added in this PR

File: slang-checks/src/main/java/com/sonarsource/slang/checks/IdenticalConditionsCheck.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  private static void checkConditions(CheckContext ctx, List<Tree> conditions) {
+    for (List<Tree> group : SyntacticEquivalence.findDuplicatedGroups(conditions)) {
+      Tree original = group.get(0);
+      group.stream().skip(1)
+        .forEach(duplicated -> ctx.reportIssue(
+          duplicated,
+          "This condition duplicates the one on line " + original.textRange().start().line() + ".",
+          new SecondaryLocation(original.textRange(), "Original")));
```
[the remark this answers is not in the document]
Comment (pynicolas) [reply slang-checks/src/main/java/com/sonarsource/slang/checks/IdenticalConditionsCheck.java]: I think it's ok.

File: slang-checks/src/main/java/com/sonarsource/slang/checks/utils/SyntacticEquivalence.java (hunk 1 above)
Comment (pynicolas) [reply slang-checks/src/main/java/com/sonarsource/slang/checks/utils/SyntacticEquivalence.java]: I think it's ok.

File: slang-checks/src/main/java/com/sonarsource/slang/checks/utils/SyntacticEquivalence.java (hunk 3 above)
Comment (pynicolas) [reply slang-checks/src/main/java/com/sonarsource/slang/checks/utils/SyntacticEquivalence.java]: I think it's ok.

File: slang-antlr/src/main/java/com/sonarsource/slang/parser/SLangConverter.java [hunk 5]
```diff
@@ -134,11 +145,55 @@ public Tree visitBlock(SLangParser.BlockContext ctx) {
       return new BlockTreeImpl(textRange(ctx.LCURLY().getSymbol(), ctx.RCURLY().getSymbol()), list(ctx.statementOrExpression()));
     }
 
+    @Override
+    public Tree visitConditional(SLangParser.ConditionalContext ctx) {
+      return visit(ctx.children.get(0));
```
[the remark this answers is not in the document]
Comment (pynicolas) [reply slang-antlr/src/main/java/com/sonarsource/slang/parser/SLangConverter.java]: fixed in 6db1484

File: slang-antlr/src/main/java/com/sonarsource/slang/parser/SLangConverter.java [hunk 6]
```diff
@@ -134,11 +145,55 @@ public Tree visitBlock(SLangParser.BlockContext ctx) {
...
+      return new IfTreeImpl(textRange(startOf(ctx.start), lastTree.textRange().end()), visit(ctx.statementOrExpression()), thenBranch, elseBranch);
+    }
+
+    @Override
+    public Tree visitMatchExpression(SLangParser.MatchExpressionContext ctx) {
+      List<MatchCaseTree> cases = new ArrayList<>();
+      for (SLangParser.MatchCaseContext matchCaseContext : ctx.matchCase()) {
+        cases.add((MatchCaseTree) visit(matchCaseContext));
+      }
+      return new MatchTreeImpl(textRange(ctx.start, ctx.RCURLY().getSymbol()), visit(ctx.statementOrExpression()), cases);
```
[the remark this answers is not in the document]
Comment (pynicolas) [reply slang-antlr/src/main/java/com/sonarsource/slang/parser/SLangConverter.java]: fixed in 6db1484

File: slang-checks/src/main/java/com/sonarsource/slang/checks/utils/SyntacticEquivalence.java (hunk 2 above)
Comment (pynicolas) [reply slang-checks/src/main/java/com/sonarsource/slang/checks/utils/SyntacticEquivalence.java]: fixed in 189d523

**Pull request merged**